### ISSUE TITLE
asyncssh 2.9.0 raises AttributeError: 'NoneType' object has no attribute 'encode' in know_hosts.py

### ISSUE THREAD

**Issue by velavokr** [ronf/asyncssh#441]
When I connect to realhost.host via `asyncssh.create_connection(asyncssh.SSHClient, host='realhost.host')` I get the following:
```
...
  File "/venv/lib/python3.8/site-packages/asyncssh/connection.py", line 7977, in create_connection
    conn = await connect(host, port, client_factory=client_factory,
  File "/venv/lib/python3.8/site-packages/asyncssh/connection.py", line 7646, in connect
    return await asyncio.wait_for(
  File "/usr/lib/python3.8/asyncio/tasks.py", line 455, in wait_for
    return await fut
  File "/venv/lib/python3.8/site-packages/asyncssh/connection.py", line 396, in _connect
    new_tunnel = await _open_tunnel(tunnel, options.passphrase)
  File "/venv/lib/python3.8/site-packages/asyncssh/connection.py", line 375, in _open_tunnel
    return await connect(host, port, username=username,
  File "/venv/lib/python3.8/site-packages/asyncssh/connection.py", line 7646, in connect
    return await asyncio.wait_for(
  File "/usr/lib/python3.8/asyncio/tasks.py", line 455, in wait_for
    return await fut
  File "/venv/lib/python3.8/site-packages/asyncssh/connection.py", line 439, in _connect
    await options.waiter
  File "/venv/lib/python3.8/site-packages/asyncssh/connection.py", line 1216, in connection_made
    self._connection_made()
  File "/venv/lib/python3.8/site-packages/asyncssh/connection.py", line 3105, in _connection_made
    self._match_known_hosts(cast(KnownHostsArg, self._known_hosts),
  File "/venv/lib/python3.8/site-packages/asyncssh/connection.py", line 1081, in _match_known_hosts
    match_known_hosts(known_hosts, host, addr, port)
  File "/venv/lib/python3.8/site-packages/asyncssh/known_hosts.py", line 374, in match_known_hosts
    known_hosts = known_hosts.match(host, addr, port)
  File "/venv/lib/python3.8/site-packages/asyncssh/known_hosts.py", line 275, in match
    x509_subjects, revoked_subjects = self._match(host, addr, port)
  File "/venv/lib/python3.8/site-packages/asyncssh/known_hosts.py", line 217, in _match
    matches += (match for (entry, match) in self._pattern_entries
  File "/venv/lib/python3.8/site-packages/asyncssh/known_hosts.py", line 218, in <genexpr>
    if entry.matches(host, addr, ip))
  File "/venv/lib/python3.8/site-packages/asyncssh/known_hosts.py", line 114, in matches
    return self._match(host) or self._match(addr)
  File "/venv/lib/python3.8/site-packages/asyncssh/known_hosts.py", line 108, in _match
    hosthash = hmac.new(self._salt, value.encode(), sha1).digest()
AttributeError: 'NoneType' object has no attribute 'encode'
```
Downgrating to 2.8.1 fixes the issue.

`.ssh/config`:
```
Host bastion.host
  ForwardAgent yes

Host realhost.host
  ProxyJump bastion.host
```

**Comment by ronf** [comment]
Thanks for the report! The problem appears to be related to matching against a hashed host entry, with either the `host` or the `addr` argument that's passed in being set to `None`. In prior releases, `None` was allowed for these arguments and the code had extra checks to skip matching when `host` or `addr` was either `None` or an empty string. However, in the process of adding type annotations, I changed this to always require a string to simplify the typing, as there was really no need to distinguish `None` from the empty string in this case.

I'm guessing there's some code path that still passes in `None`, but on an initial pass through the code I'm not seeing it yet.

I also see you're using ProxyJump here, so there's a tunnel involved. I tried to see if that might be related, but even with a ProxyJump tunnel and a hashed host entry I'm still not triggering the crash.

As a workaround, I suspect the following code change on line 114 of known_hosts.py will probably "fix" this:

Change:
```python
        return self._match(host) or self._match(addr)
```
to:
```python
        return (host and self._match(host)) or (addr and self._match(addr))
```

However, I'd prefer to get to the root cause of this problem and try and find the place where `None` is being passed in.

If you don't mind, could you add a print() in the _connection_made() function in connection.py just before line 3105 (where it calls self._match_known_hosts) to print out the values of `self._host` and `self._peer_addr`? That'll help narrow down whether the problem is with the host or the addr that's being passed to match_known_hosts().

**Comment by velavokr** [comment]
Added `print("HERE", repr(self._host), repr(self._peer_addr))`

Got the result
```
HERE 'bastion.host' '2001:db8:8:4::2'
HERE 'realhost.host' None
```

**Comment by velavokr** [comment]
With `traceback.print_stack()` just above

```
...
  File "/usr/lib/python3.8/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.8/asyncio/base_events.py", line 603, in run_until_complete
    self.run_forever()
  File "/usr/lib/python3.8/asyncio/base_events.py", line 570, in run_forever
    self._run_once()
  File "/usr/lib/python3.8/asyncio/base_events.py", line 1859, in _run_once
    handle._run()
  File "/usr/lib/python3.8/asyncio/events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
  File "/venv/lib/python3.8/site-packages/asyncssh/connection.py", line 1216, in connection_made
    self._connection_made()
  File "/venv/lib/python3.8/site-packages/asyncssh/connection.py", line 3105, in _connection_made
    traceback.print_stack()
HERE 'bastion.host' '2001:db8:8:4::2'

...
  File "/usr/lib/python3.8/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.8/asyncio/base_events.py", line 603, in run_until_complete
    self.run_forever()
  File "/usr/lib/python3.8/asyncio/base_events.py", line 570, in run_forever
    self._run_once()
  File "/usr/lib/python3.8/asyncio/base_events.py", line 1859, in _run_once
    handle._run()
  File "/usr/lib/python3.8/asyncio/events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
  File "/venv/lib/python3.8/site-packages/asyncssh/connection.py", line 7979, in create_connection
    conn = await connect(host, port, client_factory=client_factory,
  File "/venv/lib/python3.8/site-packages/asyncssh/connection.py", line 7648, in connect
    return await asyncio.wait_for(
  File "/usr/lib/python3.8/asyncio/tasks.py", line 455, in wait_for
    return await fut
  File "/venv/lib/python3.8/site-packages/asyncssh/connection.py", line 396, in _connect
    new_tunnel = await _open_tunnel(tunnel, options.passphrase)
  File "/venv/lib/python3.8/site-packages/asyncssh/connection.py", line 375, in _open_tunnel
    return await connect(host, port, username=username,
  File "/venv/lib/python3.8/site-packages/asyncssh/connection.py", line 7648, in connect
    return await asyncio.wait_for(
  File "/usr/lib/python3.8/asyncio/tasks.py", line 455, in wait_for
    return await fut
  File "/venv/lib/python3.8/site-packages/asyncssh/connection.py", line 405, in _connect
    _, tunnel_session = await new_tunnel.create_connection(
  File "/venv/lib/python3.8/site-packages/asyncssh/connection.py", line 4130, in create_connection
    session = await chan.connect(session_factory, remote_host, remote_port,
  File "/venv/lib/python3.8/site-packages/asyncssh/channel.py", line 2020, in connect
    return (await self._open_tcp(session_factory, b'direct-tcpip',
  File "/venv/lib/python3.8/site-packages/asyncssh/channel.py", line 2010, in _open_tcp
    await self._open_forward(session_factory, chantype,
  File "/venv/lib/python3.8/site-packages/asyncssh/channel.py", line 1988, in _open_forward
    self._session.connection_made(self)
  File "/venv/lib/python3.8/site-packages/asyncssh/connection.py", line 1216, in connection_made
    self._connection_made()
  File "/venv/lib/python3.8/site-packages/asyncssh/connection.py", line 3105, in _connection_made
    traceback.print_stack()
HERE 'realhost.host' None
```

**Comment by ronf** [comment]
Thanks! This confirms that it is the address and not the hostname that is the problem. I had a feeling that was going to be the case. However, there aren't many places that set the address. In fact, it looks like the only one that could potentially set it to `None` is probably in `connection_made()` in connection.py (around line 1027):

```python
        peername = cast(SockAddr, transport.get_extra_info('peername'))

        if peername: # pragma: no branch
            self._peer_addr, self._peer_port = peername[:2]
```

However, this would only happen if the "peername" info value set by asyncio was actually _set_ on the transport, but with the address portion of that tuple set to `None`, which I generally wouldn't expect.

Can you try adding a print() just after the `peername` assignment that prints its value? Also, what OS are you running this on?

**Comment by ronf** [comment]
I did a little more digging on this tonight, and found some places related to tunneling where I was still setting "peername" to be `(None, None)` when a tunnel was going over an SSH TCP session which didn't have a peer address other than one associated with the SSH connection the tunnel was carried over. I can't quite explain yet why you're seeing this crash and I'm not, but either way I think the following change is probably called for given the code no longer expects to deal with `None` for either the peer address or peer port.

```patch
diff --git a/asyncssh/channel.py b/asyncssh/channel.py
index 6092e6b..37828c8 100644
--- a/asyncssh/channel.py
+++ b/asyncssh/channel.py
@@ -2002,7 +2002,7 @@ class SSHTCPChannel(SSHForwardChannel, Generic[AnyStr]):
                         orig_port: int) -> SSHTCPSession[AnyStr]:
         """Open a TCP channel"""

-        self.set_extra_info(peername=(None, None),
+        self.set_extra_info(peername=('', 0),
                             local_peername=(orig_host, orig_port),
                             remote_peername=(host, port))

@@ -2032,7 +2032,7 @@ class SSHTCPChannel(SSHForwardChannel, Generic[AnyStr]):
                                orig_host: str, orig_port: int) -> None:
         """Set local and remote peer names for inbound connections"""

-        self.set_extra_info(peername=(None, None),
+        self.set_extra_info(peername=('', 0),
                             local_peername=(dest_host, dest_port),
                             remote_peername=(orig_host, orig_port))

@@ -2084,7 +2084,7 @@ class SSHX11Channel(SSHForwardChannel[bytes]):
         """Open an SSH X11 channel"""

         self.set_extra_info(local_peername=(orig_host, orig_port),
-                            remote_peername=(None, None))
+                            remote_peername=('', 0))

         return cast(SSHTCPSession[bytes],
                     await self._open_forward(session_factory, b'x11',
@@ -2094,7 +2094,7 @@ class SSHX11Channel(SSHForwardChannel[bytes]):
     def set_inbound_peer_names(self, orig_host: str, orig_port: int) -> None:
         """Set local and remote peer name for inbound connections"""

-        self.set_extra_info(local_peername=(None, None),
+        self.set_extra_info(local_peername=('', 0),
                             remote_peername=(orig_host, orig_port))


```

Could you try applying these changes and see if it makes a difference? In the meantime, I'll keep on trying to reproduce this here.

**Comment by ronf** [comment]
Ok - I've managed to reproduce this here. It turns out that I needed to have a hashed host entry in known_hosts which _didn't_ match the host I was trying to reach, so that it would attempt to see if the address matched. In addition, the connection had to be going though a jump host. This combination got the error you saw, and I can confirm that the above change does fix the problem.

Basically, because this is connecting over a tunnel to a hostname which only gets resolved on the upstream SSH server, there's no way to set address information for the target host, which means the known_hosts matching is only able to match on the hostname and not an IP address for that host. However, since this code was still setting the address to `None` rather than an empty string, this broke the updated known_hosts matching code which now requires this argument be a string.

I'll get a fix for this into the "develop" branch shortly and let you know here when it's available. Feel free to use the above if you like in the meantime.
